x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/LinearProgrammingPlanSolver.java

```diff
@@ -0,0 +1,341 @@
+        // Thus, we can express: y_i_j = (a_i_j * t_i) / N_j
+        //
+        // The memory constraint is then rewritten as: sum(m_i * a_i_j * t_i / N_j) <= M_j
+        // The objective function becomes: maximize sum(a_i_j * w_i_j) - L1 * sum(m_i * a_i_j * t_i / N_j).
+        // We want to choose L1 in such a way that we will always assign models to a node if there is capacity.
+        // Since w_i_j can be equal to 1 and m_i <= 1, for each node it has to be that L1 * t_i / N_j < 1.
+        // The max value for t_i is C, where C is the max number of cores across nodes. Thus, we can write L1 * C / N_j < 1.
+        // We define L1' so that L1' = L1 * C / N_j < 1. Solving for L1, we get L1 = L1' * N_j / C.
+        // Replacing L1 with L1' in the objective function we get: maximize sum(a_i_j * w_i_j) - L1' * sum(m_i * a_i_j / C).
+
+        ExpressionsBasedModel model = new ExpressionsBasedModel(
+            new Optimisation.Options().abort(new CalendarDateDuration(10, CalendarDateUnit.SECOND))
+        );
+
+        Map<Tuple<Model, Node>, Variable> allocationVars = new HashMap<>();
+
+        for (Model m : models) {
+            for (Node n : nodes) {
+                Variable allocationVar = model.addVariable("allocations_of_model_" + m.id() + "_on_node_" + n.id())
+                    .integer(false) // We relax the program to non-integer as the integer solver is much slower and can often lead to
+                                    // infeasible solutions
+                    .lower(0.0) // It is important not to set an upper bound here as it impacts memory negatively
+                    .weight(weightForAllocationVar(m, n, weights));
+                allocationVars.put(Tuple.tuple(m, n), allocationVar);
+            }
+        }
+
+        for (Model m : models) {
+            model.addExpression("allocations_of_model_" + m.id() + "_not_more_than_required")
+                .lower(m.getPreviouslyAssignedAllocations())
+                .upper(m.allocations())
+                .setLinearFactorsSimple(varsForModel(m, allocationVars));
+        }
+
+        double[] threadsPerAllocationPerModel = models.stream().mapToDouble(m -> m.threadsPerAllocation()).toArray();
+        for (Node n : nodes) {
+            model.addExpression("threads_on_node_" + n.id() + "_not_more_than_cores")
+                .upper(coresPerNode.get(n))
+                .setLinearFactors(varsForNode(n, allocationVars), Access1D.wrap(threadsPerAllocationPerModel));
+        }
+
+        for (Node n : nodes) {
+            List<Variable> allocations = new ArrayList<>();
+            List<Double> modelMemories = new ArrayList<>();
+            models.stream().filter(m -> m.currentAllocationByNodeId().containsKey(n.id()) == false).forEach(m -> {
+                allocations.add(allocationVars.get(Tuple.tuple(m, n)));
+                modelMemories.add(normalizedMemoryPerModel.get(m) * m.threadsPerAllocation() / (double) coresPerNode.get(n));
```

Comment:
Is this supposed to be
```
m_i * a_i_j * t_i / N_j
```
Or 
```
(a_i_j * t_i) / N_j
```

(If I am reading this correctly, I think its `m_i * a_i_j * t_i / N_j` as its t_i/N-j * memory and then we use `a_i_j` as the list of values...)

Overall comment to prevent confusion, it would be good in addition to the overall comment regarding the logic of the linear equations, that each line contains which constraint its building (according to the overall comment).

Reply:
It is `m_i * a_i_j * t_i / N_j`. I'll add comments for each constraint.